components/sup/src/manager/file_watcher.rs

```diff
@@ -530,49 +527,51 @@ impl<C: Callbacks, W: Watcher> FileWatcher<C, W> {
-        self.trigger_callback(CallbackType::Appeared, notify_event)
+        if self.should_trigger(notify_event) {
+            self.callbacks.file_appeared(&self.watched_file.path);
+        }
     }
 
     fn handle_file_modification(&mut self, notify_event: &Event) {
-        self.trigger_callback(CallbackType::Modified, notify_event)
+        trace!("FileWatcher::handle_file_modification\n| {:?}\n| {:?}\n",
+               notify_event,
+               self.watched_file);
+        for path_buf in notify_event.paths.iter() {
+            if Self::canonical_compare(&self.watched_file.path, path_buf) {
+                self.callbacks.file_modified(&self.watched_file.path);
+            } else {
+                self.log_ignored_path(notify_event, path_buf, None);
+            }
+        }
     }
 
     fn handle_file_disappearance(&mut self, notify_event: &Event) {
-        self.trigger_callback(CallbackType::Disappeared, notify_event)
+        if self.should_trigger(notify_event) {
+            self.callbacks.file_disappeared(&self.watched_file.path);
+        }
     }
 
-    fn trigger_callback(&mut self, callback_type: CallbackType, notify_event: &Event) {
-        trace!("FileWatcher::_trigger_callback\n| {:?}\n| {:?}\n| {:?}\n",
-               callback_type,
-               notify_event,
-               self.watched_file);
+    fn should_trigger(&mut self, notify_event: &Event) -> bool {
         for path_buf in notify_event.paths.iter() {
-            if Self::canonical_compare(&self.watched_file.path, path_buf) {
-                match callback_type {
-                    CallbackType::Appeared => {
-                        match self.manage_watch() {
-                            Ok(_) => {
-                                self.callbacks.file_appeared(&self.watched_file.path);
-                                break;
-                            }
-                            Err(e) => self.log_ignored_path(notify_event, path_buf, Some(e)),
-                        }
-                    }
-                    CallbackType::Modified => self.callbacks.file_modified(&self.watched_file.path),
-                    CallbackType::Disappeared => {
-                        match self.manage_watch() {
-                            Ok(_) => {
-                                self.callbacks.file_disappeared(&self.watched_file.path);
-                                break;
-                            }
-                            Err(e) => self.log_ignored_path(notify_event, path_buf, Some(e)),
-                        }
+            let should_signal = match self.watcher {
+                SupWatcher::Native(_) => Self::canonical_compare(&self.watched_file.path, path_buf),
+                SupWatcher::Fallback(_) => {
+                    Self::canonical_compare(&self.watched_file.path, path_buf)
+                    || Self::canonical_compare(&self.watched_file.directory, path_buf)
+                }
```

Comment:
Would the `Fallback` branch work for both cases?

Reply:
Yes. Changed.